logs/field.go

```diff
@@ -0,0 +1,211 @@
+package logs
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/trace"
+)
+
+// Field represents typed log field (a key-value pair). Adapters should determine
+// Field's type based on Type and use the corresponding getter method to retrieve
+// the value:
+//
+//	switch f.Type {
+//	case logs.IntType:
+//	    var i int = f.Int()
+//	    // handle int value
+//	case logs.StringType:
+//	    var s string = f.String()
+//	    // handle string value
+//	//...
+//	}
+//
+// Getter methods must not be called on fields with wrong Type (e.g. calling String()
+// on fields with Type != StringType).
+type Field struct {
+	Type FieldType
+	Key  string
+
+	vint int64
+	vstr string
+	vany interface{}
+}
+
+func String(key string, value string) Field {
+	return Field{
+		Type: StringType,
+		Key:  key,
+		vstr: value,
+	}
+}
+
+func (f Field) String() string {
+	return f.vstr
+}
+
+func Int(key string, value int) Field {
+	return Field{
+		Type: IntType,
+		Key:  key,
+		vint: int64(value),
+	}
+}
+
+func (f Field) Int() int {
+	return int(f.vint)
+}
+
+func Bool(key string, value bool) Field {
+	var vint int64
+	if value {
+		vint = 1
+	} else {
+		vint = 0
+	}
+	return Field{
+		Type: BoolType,
+		Key:  key,
+		vint: vint,
+	}
+}
+
+func (f Field) Bool() bool {
+	return f.vint != 0
+}
+
+func Duration(key string, value time.Duration) Field {
+	return Field{
+		Type: DurationType,
+		Key:  key,
+		vint: value.Nanoseconds(),
+	}
+}
+
+func (f Field) Duration() time.Duration {
+	return time.Nanosecond * time.Duration(f.vint)
+}
+
+func Strings(key string, value []string) Field {
+	return Field{
+		Type: StringsType,
+		Key:  key,
+		vany: value,
+	}
+}
+
+func (f Field) Strings() []string {
+	return f.vany.([]string)
+}
+
+// NamedError adds error field. If value is nil, resulting Field will be
+// of NilType instead of ErrorType.
+func NamedError(key string, value error) Field {
+	if value == nil {
+		return nilField(key)
+	}
+	return Field{
+		Type: ErrorType,
+		Key:  key,
+		vany: value,
+	}
+}
+
+// Error is the same as NamedError("error", value)
+func Error(value error) Field {
+	return NamedError("error", value)
+}
+
+func (f Field) Error() error {
+	return f.vany.(error)
+}
+
+// Any adds untyped field. If value is nil, resulting Field will be
+// of NilType instead of AnyType.
+func Any(key string, value interface{}) Field {
+	if value == nil {
+		return nilField(key)
+	}
+	return Field{
+		Type: AnyType,
+		Key:  key,
+		vany: value,
+	}
+}
+
+func (f Field) Any() interface{} {
+	return f.vany
+}
+
+func nilField(key string) Field {
+	return Field{
+		Type: NilType,
+		Key:  key,
+	}
+}
+
+func Endpoints(key string, value []trace.EndpointInfo) Field {
```

Comment:
I think we should no add internal structures as field types.

If somebody with adapter X will "optimize" output endpoints, then we will add new field for our trace.EndpointInfo: we will no see new field in the logs from adapter X.

All convert from internal stuctures to plain types will in our internal logger.

Reply:
The problem here is that `[]trace.EndpointInfo` is essentially `[]fmt.Stringer`, but it cannot be converted to that type without reallocations/reflection. So adding internal (exported, to be exact) types as FieldType is just a way to make logger as zero-cost as possible.

Zap [solves](https://pkg.go.dev/go.uber.org/zap#Stringers) this by using generics, but it is not an option if we want to support users with older Go versions.

Reply:
We can create type endpoints
```
type Endpoints []trace.EndpointInfo
```

local to tracer package and implement stringer for Endpoints (not for every endpoints). Convert from []trace.EndpointInfo to Endpoints is cheap and no need allocations.

Reply:
It is a solution, but it makes slice of endpoints to always be one field with no internal structure, whereas now adapters can make it a nested array of fields if they support it. If we marshal endpoints to text, we need to choose how to marshal it and take away this choice from user/adapter/etc. For example, we can stringify `[]trace.EndpointInfo` as JSON array, but what if someone wants to use some other markup language for log messages?

We could also do something like
```go
type Marshaler interface {
    json.Marshaler
    xml.Marshaler
    ...
}

func MarshalerField(key string, value Marshaler) Field {...}
```
But it just replaces one **fixed** format with several **fixed** formats.

Reply:
Replaced internal types with `fmt.Stringer`.